bcachefs.8
.Dd November 17, 2023
.Dt BCACHEFS 8 SMM
.Os
.Sh NAME
.Nm bcachefs
.Nd manage bcachefs filesystems/devices
.Sh SYNOPSIS
.Nm
.Ar command
.Op Ar options
.Op Ar arguments
.Sh DESCRIPTION
The
.Nm
utility supports the following subcommands,
which are documented in detail below:
.Ss Superblock commands
.Bl -tag -width 18n -compact
.It Ic format
Format one or a list of devices with bcachefs data structures.
.It Ic show-super
Dump superblock information to stdout.
.It Ic set-fs-option
Set a filesystem option
.El
.Ss Mount commands
.Bl -tag -width 18n -compact
.It Ic mount
Mount a filesystem.
.El
.Ss Repair commands
.Bl -tag -width 18n -compact
.It Ic fsck
Check an existing filesystem for errors.
.El
.Ss Commands for managing a running filesystem
.Bl -tag -width 18n -compact
.It Ic fs usage
Show disk usage
.El
.Ss Commands for managing devices within a running filesystem
.Bl -tag -width 22n -compact
.It Ic device add
Add a new device to an existing filesystem
.It Ic device remove
Remove a device from an existing filesystem
.It Ic device online
Re-add an existing member to a filesystem
.It Ic device offline
Take a device offline, without removing it
.It Ic device evacuate
Migrate data off of a specific device
.It Ic device set-state
Mark a device as failed
.It Ic device resize
Resize filesystem on a device
.It Ic device resize-journal
Resize journal on a device
.El
.Ss Commands for managing subvolumes and snapshots
.Bl -tag -width 18n -compact
.It Ic subvolume create
Create a new subvolume
.It Ic subvolume delete
Delete an existing subvolume
.It Ic subvolume snapshot
Create a snapshot
.El
.Ss Commands for managing filesystem data
.Bl -tag -width 18n -compact
.It Ic data rereplicate
Rereplicate degraded data
.It Ic data job
Kick off low level data jobs
.El
.Ss Commands for encryption
.Bl -tag -width 18n -compact
.It Ic unlock
Unlock an encrypted filesystem prior to running/mounting
.It Ic set-passphrase
Change passphrase on an existing (unmounted) filesystem
.It Ic remove-passphrase
Remove passphrase on an existing (unmounted) filesystem
.El
.Ss Commands for migration
.Bl -tag -width 18n -compact
.It Ic migrate
Migrate an existing filesystem to bcachefs, in place
.It Ic migrate-superblock
Add default superblock, after bcachefs migrate
.El
.Ss Commands for operating on files in a bcachefs filesystem
.Bl -tag -width 18n -compact
.It Ic set-file-option
Set various per file attributes
.El
.Ss Commands for debugging
.Bl -tag -width 18n -compact
.It Ic dump
Dump filesystem metadata to a qcow2 image
.It Ic list
List filesystem metadata in textual form
.It Ic list_journal
List contents of journal
.El
.Ss FUSE commands
.Bl -tag -width 18n -compact
.It Ic fusemount Mount a filesystem via FUSE
.El
.Ss Miscellaneous commands
.Bl -tag -width 18n -compact
.It Ic version
Display the version of the invoked bcachefs tool
.It Ic completions
Generate shell completions
.El
.Sh Superblock commands
.Bl -tag -width Ds
.It Nm Ic format Oo Ar options Oc Ar devices\ ...
Format one or a list of devices with bcachefs data structures.
You need to do this before you create a volume.
.Pp
Device specific options must come before corresponding devices, e.g.
.Dl bcachefs format --label=ssd /dev/sda --label=hdd /dev/sdb
.Bl -tag -width Ds
.It Fl -block_size Ns = Ns Ar size
block size, in bytes (e.g. 4k)
.It Fl -btree_node_size Ns = Ns Ar size
Btree node size, default 256k
.It Fl -errors Ns = Ns ( Cm continue | ro | panic )
Action to take on filesystem error
.It Fl -data_replicas Ns = Ns Ar number
Number of data replicas
.It Fl -metadata_replicas Ns = Ns Ar number
Number of metadata replicas
.It Fl -data_replicas_required Ns = Ns Ar number

.It Fl -metadata_replicas_required Ns = Ns Ar number

.It Fl -encoded_extent_max Ns = Ns Ar size
Maximum size of checksummed/compressed extents
.It Fl -metadata_checksum Ns = Ns ( Cm none | crc32c | crc64 | xxhash )
Set metadata checksum type (default:
.Cm crc32c ) .
.It Fl -data_checksum Ns = Ns ( Cm none | crc32c | crc64 | xxhash )
Set data checksum type (default:
.Cm crc32c ) .
.It Fl -compression Ns = Ns ( Cm none | lz4 | gzip | zstd )
Set compression type (default:
.Cm none ) .
.It Fl -background_compression Ns = Ns ( Cm none | lz4 | gzip | zstd )

.It Fl -str_hash Ns = Ns ( Cm crc32c | crc64 | siphash )
Hash function for directory entries and xattrs
.It Fl -metadata_target Ns = Ns Ar target
Device or label for metadata writes
.It Fl -foreground_target Ns = Ns Ar target
Device or label for foreground writes
.It Fl -background_target Ns = Ns Ar target
Device or label to move data to in the background
.It Fl -promote_target Ns = Ns Ar target
Device or label to promote data to on read
.It Fl -erasure_code
Enable erasure coding (DO NOT USE YET)
.It Fl -inodes_32bit
Constrain inode numbers to 32 bits
.It Fl -shared_inode_numbers
Shared new inode numbers by CPU id
.It Fl -inodes_use_key_cache
Use the btree key cache for the inodes btree
.It Fl -gc_reserve_percent Ns = Ns Ar percentage
Percentage of disk space to reserve for copygc
.It Fl -gc_reserve_bytes Ns = Ns Ar percentage
Amount of disk space to reserve for copygc
.sp
This takes precedence over
.Cm gc_reserve_percent
if set
.It Fl -root_reserve_percent Ns = Ns Ar percentage
Percentage of disk space to reserve for superuser
.It Fl -wide_macs
Store full 128bits of cryptographic MACS, instead of 80
.It Fl -acl
Enable POSIX acls
.It Fl -usrquota
Enable user quotas
.It Fl -grpquota
Enable group quotas
.It Fl prjquota
Enable project quotas
.It Fl -journal_transaction_names
Log transaction function names in journal
.It Fl -nocow
Nocow mode: Writes will be done in place when possible.
.sp
Snapshots and reflink will still caused writes to be COW.
.sp
This flag implicitly disables data checksumming, compression and
encryption.
.It Fl -replicas Ns = Ns Ar number
Sets both data and metadata replicas
.It Fl -encrypted
Enable whole filesystem encryption (chacha20/poly1305);
passphrase will be prompted for.
.It Fl -no_passphrase
Don't encrypt master encryption key
.It Fl L , Fl -fs_label Ns = Ns Ar label
Create the filesystem with the specified
.Ar label
.It Fl U , -uuid Ns = Ns Ar uuid
Create the filesystem with the specified
.Ar uuid
.It Fl -superblock_size Ns = Ns Ar size

.El
.Pp
Device specific options:
.Bl -tag -width Ds
.It Fl -discard
Enable discard/TRIM support
.It Fl -fs_size Ns = Ns Ar size
Create the filesystem using
.Ar size
bytes on the subsequent device.
.It Fl -bucket Ns = Ns Ar size
Specifies the bucket size;
must be greater than the btree node size
.It Fl -durability Ns = Ns Ar n
Data written to this device will be considered
to have already been replicated
.Ar n
times
.It Fl l , Fl -label
Disk label
.It Fl f , Fl -force
Force the filesystem to be created,
even if the device already contains a filesystem.
.It Fl q , Fl -quiet
Only print errors
.It Fl v , Fl -verbose
Verbose filesystem initialization
.El
.It Nm Ic show-super Oo Ar options Oc Ar device
Dump superblock information to stdout.
.Bl -tag -width Ds
.It Fl f , Fl -fields Ns = Ns Ar fields
List of sections to print
.It Fl l , Fl -layout
Print superblock layout
.El
.It Nm Ic set-fs-option Oo Ar options Oc Ar device
.Bl -tag -width Ds
.It Fl -errors Ns = Ns ( Cm continue | ro | panic )
Action to take on filesystem error
.It Fl -metadata_replicas Ns = Ns Ar number
Number of metadata replicas
.It Fl -data_replicas Ns = Ns Ar number
Number of data replicas
.It Fl -metadata_replicas_required Ns = Ns Ar number

.It Fl -data_replicas_required Ns = Ns Ar number

.It Fl -metadata_checksum Ns = Ns ( Cm none | crc32c | crc64 | xxhash )
Set metadata checksum type (default:
.Cm crc32c ) .
.It Fl -data_checksum Ns = Ns ( Cm none | crc32c | crc64 | xxhash )
Set data checksum type (default:
.Cm crc32c ) .
.It Fl -compression Ns = Ns ( Cm none | lz4 | gzip | zstd )
Set compression type (default:
.Cm none ) .
.It Fl -background_compression Ns = Ns ( Cm none | lz4 | gzip | zstd )

.It Fl -str_hash Ns = Ns ( Cm crc32c | crc64 | siphash )
Hash function for directory entries and xattrs
.It Fl -metadata_target Ns = Ns Ar target
Device or label for metadata writes
.It Fl -foreground_target Ns = Ns Ar target
Device or label for foreground writes
.It Fl -background_target Ns = Ns Ar target
Device or label to move data to in the background
.It Fl -promote_target Ns = Ns Ar target
Device or label to promote data to on read
.It Fl -erasure_code
Enable erasure coding (DO NOT USE YET)
.It Fl -inodes_32bit
Constrain inode numbers to 32 bits
.It Fl -shared_inode_numbers
Shared new inode numbers by CPU id
.It Fl -inodes_use_key_cache
Use the btree key cache for the inodes btree
.It Fl -gc_reserve_percent Ns = Ns Ar percentage
Percentage of disk space to reserve for copygc
.It Fl -gc_reserve_bytes Ns = Ns Ar percentage
Amount of disk space to reserve for copygc
.sp
This takes precedence over
.Cm gc_reserve_percent
if set
.It Fl -root_reserve_percent Ns = Ns Ar percentage
Percentage of disk space to reserve for superuser
.It Fl -wide_macs
Store full 128bits of cryptographic MACS, instead of 80
.It Fl -acl
Enable POSIX acls
.It Fl -usrquota
Enable user quotas
.It Fl -grpquota
Enable group quotas
.It Fl -prjquota
Enable project quotas
.It Fl -degraded
Allow mounting in degraded mode
.It Fl -very_degraded
Allow mounting in when data will be missing
.It Fl -discard
Enable discard/TRIM support
.It Fl -verbose
Extra debugging information during mount/recovery
.It Fl -journal_flush_delay Ns = Ns Ar ms
Delay in milliseconds before automatic journal commits
.It Fl -journal_flush_disabled
Disable journal flush on sync/fsync
.sp
If enabled, writes can be lost, but only since the
last journal write (default 1 second)
.It Fl -journal_reclaim_delay Ns = Ns Ar ms
Delay in milliseconds before automatic journal reclaim
.It Fl -move_bytes_in_flight Ns = Ns Ar bytes
Maximum Amount of IO to keep in flight by the move path
.It Fl -move_ios_in_flight Ns = Ns Ar number
Maximum number of IOs to keep in flight by the move path
.It Fl -fsck
Run fsck on mount
.It Fl -fix_errors Ns = Ns Ar error
Fix errors during fsck without asking
.It Fl -ratelimit_errors
Ratelimit error messages during fsck
.It Fl -nochanges
Super read only mode - no writes at all will be issued,
even if we have to replay the journal
.It Fl -norecovery
Don't replay the journal
.It Fl -journal_transaction_names
Log transaction function names in journal
.It Fl -noexcl
Don't open device in exclusive mode
.It Fl -direct_io
Use O_DIRECT (userspace only)
.It Fl -sb Ns = Ns Ar offset
Sector offset of superblock
.It Fl -reconstruct_alloc
Reconstruct alloc btree
.It Fl -version_upgrade Ns = Ns ( Cm compatible | incompatible | none )
Set superblock to latest version, allowing any new features
to be used
.It Fl -nocow
Nocow mode: Writes will be done in place when possible.
.sp
Snapshots and reflink will still caused writes to be COW.
.sp
This flag implicitly disables data checksumming, compression and
encryption.
.It Fl -nocow_enabled
Enable nocow mode: enables runtime locking in data
move path needed if nocow will ever be in use
.It Fl -no_data_io
Skip submit_bio() for data reads and writes,
for performance testing purposes
.El
.El
.Sh Mount commands
.Bl -tag -width Ds
.It Nm Ic mount Oo Ar options Oc Ar device mountpoint
Mount a filesystem. The
.Ar device
can be a device, a colon-separated list of devices, or UUID=<UUID>. The
.Ar mountpoint
is the path where the filesystem should be mounted. If not set, then the filesystem won't actually be mounted
but all steps preceding mounting the filesystem (e.g. asking for passphrase) will still be performed.
.Pp the options are as follows:
.Bl -tag -width Ds
.It Fl o Ar options
Mount options provided as a comma-separated list. See user guide for complete list.
.Bl -tag -width Ds -compact
.It Cm degraded
Allow mounting with data degraded
.It Cm verbose
Extra debugging info during mount/recovery
.It Cm fsck
Run fsck during mount
.It Cm fix_errors
Fix errors without asking during fsck
.It Cm read_only
Mount in read only mode
.It Cm version_upgrade
.El
.It Fl k , Fl -key-location Ns = Ns ( Cm fail | wait | ask )
Where the password would be loaded from. (default:
.Cm ask ) .
.Bl -tag -width Ds -compact
.It Cm fail
don't ask for password, fail if filesystem is encrypted.
.It Cm wait
wait for password to become available before mounting.
.It Cm ask
prompt the user for password.
.El
.It Fl c , Fl -colorize Ns = Ns ( Cm true | false )
Force color on/off. Default: auto-detect TTY
.It Fl v
Be verbose. Can be specified more than once.
.El
.El
.Sh Repair commands
.Bl -tag -width Ds
.It Nm Ic fsck Oo Ar options Oc Ar devices\ ...
Check an existing filesystem for errors.
.Bl -tag -width Ds
.It Fl p
Automatic repair (no questions)
.It Fl n
Don't repair, only check for errors
.It Fl y
Assume "yes" to all questions
.It Fl f
Force checking even if filesystem is marked clean
.It Fl r , Fl -ratelimit_errors
Don't display more than 10 errors of a given type
.It Fl R , Fl -reconstruct_alloc
Reconstruct the alloc btree
.It Fl v
Be verbose
.El
.El
.Sh Commands for managing a running filesystem
.Bl -tag -width Ds
.It Nm Ic fs Ic usage Oo Ar options Oc Op Ar filesystem
Show disk usage.
.Bl -tag -width Ds
.It Fl h , Fl -human-readable
Print human readable sizes.
.El
.El
.Sh Commands for managing devices within a running filesystem
.Bl -tag -width Ds
.It Nm Ic device Ic add Oo Ar options Oc Ar device
Add a device to an existing filesystem.
.Bl -tag -width Ds
.It Fl -fs_size Ns = Ns Ar size
Size of filesystem on device
.It Fl -bucket Ns = Ns Ar size
Set bucket size
.It Fl -discard
Enable discards
.It Fl l , Fl -label Ns = Ns Ar label
Disk label
.It Fl f , Fl -force
Use device even if it appears to already be formatted
.El
.It Nm Ic device Ic remove Oo Ar options Oc Ar device
Remove a device from a filesystem
.Bl -tag -width Ds
.It Fl f , Fl -force
Force removal, even if some data couldn't be migrated
.It Fl F , Fl -force-metadata
Force removal, even if some metadata couldn't be migrated
.El
.It Nm Ic device Ic online Ar device
Re-add a device to a running filesystem
.It Nm Ic device Ic offline Ar device
Take a device offline, without removing it
.Bl -tag -width Ds
.It Fl f , Fl -force
Force, if data redundancy will be degraded
.El
.It Nm Ic device Ic evacuate Ar device
Move data off of a given device
.It Nm Ic device Ic set-state Oo Ar options Oc Ar new-state Ar device
.Bl -tag -width Ds
.It Ar  new-state Ns = Ns ( Ar rw | ro | failed | spare )
.It Fl f , Fl -force
Force, if data redundancy will be degraded
.It Fl -force-if-data-lost
Force, if data will be lost
.It Fl o , Fl -offline
Set state of an offline device
.El
.It Nm Ic device Ic resize Ar device Op Ar size
Resize filesystem on a device
.It Nm Ic device Ic resize-journal Ar device Op Ar size
Resize journal on a device
.El
.Sh Commands for managing subvolumes and snapshots
.Bl -tag -width Ds
.It Ic subvolume create Oo Ar options Oc Ar path
Create a new subvolume
.It Ic subvolume delete Oo Ar options Oc Ar path
Delete an existing subvolume
.It Ic subvolume snapshot Oo Ar options Oc Ar source dest
Create a snapshot of
.Ar source
at
.Ar dest .
If specified,
.Ar source
must be a subvolume;
if not specified the snapshot will be of the subvolume containing
.Ar dest .
.Bl -tag -width Ds
.It Fl r
Make snapshot read-only
.El
.El
.Sh Commands for managing filesystem data
.Bl -tag -width Ds
.It Nm Ic data Ic rereplicate Ar filesystem
Walks existing data in a filesystem,
writing additional copies of any degraded data.
.It Nm Ic data Ic job Ar job filesystem
Kick off a data job and report progress
.sp
.Ar job
is one of (
.Cm scrub | rereplicate | migrate | rewrite_old_nodes
)
.Bl -tag -width Ds
.It Fl b Ar btree
Btree to operate on
.It Fl s Ar inode Ns Cm \&: Ns Ar offset
Start position
.It Fl e Ar inode Ns Cm \&: Ns Ar offset
End position
.El
.El
.Sh Commands for encryption
.Bl -tag -width Ds
.It Nm Ic unlock Ar device
Unlock an encrypted filesystem prior to running/mounting.
.Bl -tag -width Ds
.It Fl c
Check if a device is encrypted
.It Fl k Ns = Ns ( Cm session | user | user_session )
Keyring to add to (default:
.Cm user )
.El
.It Nm Ic set-passphrase Ar devices\ ...

debian/changelog
bcachefs-tools (1:1.9.1-1) unstable; urgency=medium

  * New upstream release
  * New build-depency: librust-strum-dev
  * Update copyright file

 -- Jonathan Carter <jcc@debian.org>  Thu, 20 Jun 2024 17:28:12 +0200

bcachefs-tools (1:1.7.0-1) unstable; urgency=medium

  * Upload to unstable (Closes: #1066929)
  * Bump epoch
    - Ack from pochu:
      https://lists.debian.org/msgid-search/1c5f86c9-1525-4d44-996f-3d4eed1e64d6@debian.org

 -- Jonathan Carter <jcc@debian.org>  Wed, 15 May 2024 12:04:38 +0200

bcachefs-tools (24+really1.7.0-1~exp1) experimental; urgency=medium

  [ Steinar H. Gunderson ]
  * New upstream release.
    * Remove the do-not-install-mount-symlink patch.
    * Add epoch to deal with new upstream versioning scheme,
      and disable pristine-tar option in gbp.conf, since there is no
      pristine-tar branch. (Closes: #1054620)
    * revert-bindgen-changes.patch: New patch, reinstates upstream
      hack that makes bcachefs-tools work with bindgen older than 0.69.4
      (which isn't in Debian yet).
  * Build-depend on systemd-dev, since udev.pc has moved there.
  * Enable Rust parts, including mount.bcachefs. (Closes: #1060256)
    * Include the right .mk flags to get DEB_HOST_RUST_TYPE etc., and
      export that (needed for Debian's cargo wrapper).
    * Override CARGO, and set CARGO_HOME during installation.
    * Run prepare-debian in dh_auto_configure, and clean up the vendor dir
      in dh_auto_clean.
    * Remove Cargo.lock before the build, so that we can use Debian's versions
      of all Rust crates. Also remove it in dh_auto_clean, so that it does not
      take a diff.
    * revert-bindgen-changes.patch: New patch, loosen required versions
      of Rust crates
      - errno (0.2 -> >= 0.2, < 1; Debian has 0.4)
      - udev (0.7 -> >= 0.7, < 1; Debian has 0.8)
      - memoffset (0.8.0 -> 0.6; Debian has 0.6.5, so a downgrade)
      - paste (1.0.11 -> 1.0; Debian has 1.0.8, so a downgrade)
      - bindgen (0.69.4 -> 0.66; see revert-bindgen-changes.patch above)
    * Add build-dependency on all relevant Rust crates, and on python3:native
      due to use of the Cargo wrapper.
  * Build-Depend on pkgconf instead of pkg-config (Lintian warning).
  * Add debhelper tokens to postinst and postrm scripts (Lintian warning).
  * Depend on python3:any due to bcachefsck_all being a Python script
    (Lintian warning).

  [ Jonathan Carter ]
  * Upload to experimental
  * Add librust-chrono-dev, librust-getset-dev and
    librust-gag-dev to build-depends
  * Recreate dependencies patch to match new upstream source
  * Update standards version to 4.7.0
  * Update above mentioned revert-bindgen-changes.patch

 -- Jonathan Carter <jcc@debian.org>  Thu, 25 Apr 2024 13:53:55 +0200

bcachefs-tools (24+really1.3.4-2) unstable; urgency=medium

  * Upload to unstable

 -- Jonathan Carter <jcc@debian.org>  Wed, 03 Jan 2024 17:06:16 +0200

bcachefs-tools (24+really1.3.4-2~exp1) experimental; urgency=medium

  [ Chris Hofstaedtler ]
  * Non-maintainer upload.
  * Install files into /usr instead of /. (Closes: #1059373)

  [ Jonathan Carter ]
  * Do not install bcachefs.mount symlink (Closes: #1057295)

 -- Jonathan Carter <jcc@debian.org>  Wed, 27 Dec 2023 19:22:06 +0200

bcachefs-tools (24+really1.3.4-1) unstable; urgency=medium

  * New upstream release

 -- Jonathan Carter <jcc@debian.org>  Tue, 21 Nov 2023 17:26:13 +0200

bcachefs-tools (24+really1.2-1) unstable; urgency=medium

  * New upstream release (Closes: #1054613)
  * Disable new rust build
    (dependencies currently unavailable in Debian, see debian/README.todo)
  * Remove unneeded override_dh_auto_clean from debian/rules
    (Closes: #1043654)

 -- Jonathan Carter <jcc@debian.org>  Thu, 28 Sep 2023 19:54:47 +0200

bcachefs-tools (24-1) unstable; urgency=medium

  * New upstream release

 -- Jonathan Carter <jcc@debian.org>  Tue, 29 Nov 2022 09:40:27 +0200

bcachefs-tools (23-1) unstable; urgency=medium

  * New upstream release
  * Update standards version to 4.6.1

 -- Jonathan Carter <jcc@debian.org>  Mon, 31 Oct 2022 11:45:25 +0200

bcachefs-tools (0.1+git20220216.a1e928a-1) unstable; urgency=medium

  * New upstream snapshot
  * Grab patch from Ubuntu to reduce memory on amd64 builders
    (http://launchpadlibrarian.net/580140160/bcachefs-tools_0.1+git20210805.6c42566-2_0.1+git20210805.6c42566-2ubuntu1.diff.gz)
  * Update copyright years

 -- Jonathan Carter <jcc@debian.org>  Wed, 16 Feb 2022 14:42:20 +0200

bcachefs-tools (0.1+git20210805.6c42566-2) unstable; urgency=medium

  * Remove valgrind as build-dependency, seems unneeded unless
    doing debug work and is not available on all architectures.

 -- Jonathan Carter <jcc@debian.org>  Fri, 03 Sep 2021 16:07:11 +0200

bcachefs-tools (0.1+git20210805.6c42566-1) unstable; urgency=medium

  * New upstream snapshot
  * Update standards version to 4.6.0
  * Add python3-docutils as dependency
  * Do not run tests at package build time

 -- Jonathan Carter <jcc@debian.org>  Tue, 06 Apr 2021 15:11:27 +0200

bcachefs-tools (0.1+git20201025.742dbbdb-1) unstable; urgency=medium

  * New upstream snapshot

 -- Jonathan Carter <jcc@debian.org>  Mon, 26 Oct 2020 08:45:37 +0200

bcachefs-tools (0.1+git20201017.8a4408-1~exp1) unstable; urgency=medium

  * New upstream release
  * Bump debhelper-compat to 13
  * Update standards version to 4.5.0
  * Declare Rules-Requires-Root: no
  * debian/copyright:
    - Update copyright years
    - Update copyright owners
    - Add intel and ccan copyright
  * Add build-dep on libudev-dev, python3-pytest  and valgrind

 -- Jonathan Carter <jcc@debian.org>  Sun, 18 Oct 2020 17:29:27 +0200

bcachefs-tools (0.1+git20190829.aa2a42b-1~exp1) experimental; urgency=medium

  * Initial Release (Closes: #935178)

 -- Jonathan Carter <jcc@debian.org>  Mon, 16 Sep 2019 10:36:04 +0000


c_src/bcachefs.c
/*
 * Authors: Kent Overstreet <kent.overstreet@gmail.com>
 *	    Gabriel de Perthuis <g2p.code@gmail.com>
 *	    Jacob Malevich <jam@datera.io>
 *
 * GPLv2
 */

#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <inttypes.h>
#include <limits.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <raid/raid.h>

#include "cmds.h"

void bcachefs_usage(void)
{
	puts("bcachefs - tool for managing bcachefs filesystems\n"
	     "usage: bcachefs <command> [<args>]\n"
	     "\n"
	     "Superblock commands:\n"
	     "  format                   Format a new filesystem\n"
	     "  show-super               Dump superblock information to stdout\n"
	     "  set-fs-option            Set a filesystem option\n"
	     "  reset-counters           Reset all counters on an unmounted device\n"
	     "\n"
	     "Mount:\n"
	     "  mount                    Mount a filesystem\n"
	     "\n"
	     "Repair:\n"
	     "  fsck                     Check an existing filesystem for errors\n"
	     "\n"
#if 0
	     "Startup/shutdown, assembly of multi device filesystems:\n"
	     "  assemble                 Assemble an existing multi device filesystem\n"
	     "  incremental              Incrementally assemble an existing multi device filesystem\n"
	     "  run                      Start a partially assembled filesystem\n"
	     "  stop	                 Stop a running filesystem\n"
	     "\n"
#endif
	     "Commands for managing a running filesystem:\n"
	     "  fs usage                 Show disk usage\n"
	     "\n"
	     "Commands for managing devices within a running filesystem:\n"
	     "  device add               Add a new device to an existing filesystem\n"
	     "  device remove            Remove a device from an existing filesystem\n"
	     "  device online            Re-add an existing member to a filesystem\n"
	     "  device offline           Take a device offline, without removing it\n"
	     "  device evacuate          Migrate data off of a specific device\n"
	     "  device set-state         Mark a device as failed\n"
	     "  device resize            Resize filesystem on a device\n"
	     "  device resize-journal    Resize journal on a device\n"
	     "\n"
	     "Commands for managing subvolumes and snapshots:\n"
	     "  subvolume create         Create a new subvolume\n"
	     "  subvolume delete         Delete an existing subvolume\n"
	     "  subvolume snapshot       Create a snapshot\n"
	     "\n"
	     "Commands for managing filesystem data:\n"
	     "  data rereplicate         Rereplicate degraded data\n"
	     "  data scrub               Verify checksums and correct errors, if possible\n"
	     "  data job                 Kick off low level data jobs\n"
	     "\n"
	     "Encryption:\n"
	     "  unlock                   Unlock an encrypted filesystem prior to running/mounting\n"
	     "  set-passphrase           Change passphrase on an existing (unmounted) filesystem\n"
	     "  remove-passphrase        Remove passphrase on an existing (unmounted) filesystem\n"
	     "\n"
	     "Migrate:\n"
	     "  migrate                  Migrate an existing filesystem to bcachefs, in place\n"
	     "  migrate-superblock       Add default superblock, after bcachefs migrate\n"
	     "\n"
	     "Commands for operating on files in a bcachefs filesystem:\n"
	     "  set-file-option          Set various attributes on files or directories\n"
	     "\n"
	     "Debug:\n"
	     "These commands work on offline, unmounted filesystems\n"
	     "  dump                     Dump filesystem metadata to a qcow2 image\n"
	     "  list                     List filesystem metadata in textual form\n"
	     "  list_journal             List contents of journal\n"
	     "\n"
	     "FUSE:\n"
	     "  fusemount                Mount a filesystem via FUSE\n"
	     "\n"
	     "Miscellaneous:\n"
	     "  completions              Generate shell completions\n"
	     "  version                  Display the version of the invoked bcachefs tool\n");
}

static char *pop_cmd(int *argc, char *argv[])
{
	char *cmd = argv[1];
	if (!(*argc < 2))
		memmove(&argv[1], &argv[2], (*argc - 2) * sizeof(argv[0]));
	(*argc)--;
	argv[*argc] = NULL;

	return cmd;
}

int fs_cmds(int argc, char *argv[])
{
	char *cmd = pop_cmd(&argc, argv);

	if (argc < 1) {
		bcachefs_usage();
		exit(EXIT_FAILURE);
	}
	if (!strcmp(cmd, "usage"))
		return cmd_fs_usage(argc, argv);

	return 0;
}

int device_cmds(int argc, char *argv[])
{
	char *cmd = pop_cmd(&argc, argv);

	if (argc < 1)
		return device_usage();
	if (!strcmp(cmd, "add"))
		return cmd_device_add(argc, argv);
	if (!strcmp(cmd, "remove"))
		return cmd_device_remove(argc, argv);
	if (!strcmp(cmd, "online"))
		return cmd_device_online(argc, argv);
	if (!strcmp(cmd, "offline"))
		return cmd_device_offline(argc, argv);
	if (!strcmp(cmd, "evacuate"))
		return cmd_device_evacuate(argc, argv);
	if (!strcmp(cmd, "set-state"))
		return cmd_device_set_state(argc, argv);
	if (!strcmp(cmd, "resize"))
		return cmd_device_resize(argc, argv);
	if (!strcmp(cmd, "resize-journal"))
		return cmd_device_resize_journal(argc, argv);

	return 0;
}

int data_cmds(int argc, char *argv[])
{
	char *cmd = pop_cmd(&argc, argv);

	if (argc < 1)
		return data_usage();
	if (!strcmp(cmd, "rereplicate"))
		return cmd_data_rereplicate(argc, argv);
	if (!strcmp(cmd, "scrub"))
		return cmd_data_scrub(argc, argv);
	if (!strcmp(cmd, "job"))
		return cmd_data_job(argc, argv);

	return 0;
}

int subvolume_cmds(int argc, char *argv[])
{
	char *cmd = pop_cmd(&argc, argv);
	if (argc < 1)
		return subvolume_usage();
	if (!strcmp(cmd, "create") || !strcmp(cmd, "new"))
		return cmd_subvolume_create(argc, argv);
	if (!strcmp(cmd, "delete") || !strcmp(cmd, "del"))
		return cmd_subvolume_delete(argc, argv);
	if (!strcmp(cmd, "snapshot") || !strcmp(cmd, "snap"))
		return cmd_subvolume_snapshot(argc, argv);

	return 0;
}

int main(int argc, char *argv[])
{
	raid_init();

    setvbuf(stdout, NULL, _IOLBF, 0);

    char *full_cmd = argv[0];

	/* Are we being called via a symlink? */

	if (strstr(full_cmd, "mkfs"))
		return cmd_format(argc, argv);

	if (strstr(full_cmd, "fsck"))
		return cmd_fsck(argc, argv);

#ifdef BCACHEFS_FUSE
	if (strstr(full_cmd, "mount.fuse"))
		return cmd_fusemount(argc, argv);
#endif

	if (strstr(full_cmd, "mount"))
		return cmd_mount(argc, argv);

	char *cmd = pop_cmd(&argc, argv);
	if (!cmd) {
		puts("missing command\n");
		goto usage;
	}

	/* these subcommands display usage when argc < 2 */
	if (!strcmp(cmd, "device"))
		return device_cmds(argc, argv);
	if (!strcmp(cmd, "fs"))
		return fs_cmds(argc, argv);
	if (!strcmp(cmd, "data"))
		return data_cmds(argc, argv);
	if (!strcmp(cmd, "subvolume"))
		return subvolume_cmds(argc, argv);
	if (!strcmp(cmd, "format"))
		return cmd_format(argc, argv);
	if (!strcmp(cmd, "fsck"))
		return cmd_fsck(argc, argv);
	if (!strcmp(cmd, "version"))
		return cmd_version(argc, argv);
	if (!strcmp(cmd, "show-super"))
		return cmd_show_super(argc, argv);
	if (!strcmp(cmd, "set-option"))
		return cmd_set_option(argc, argv);
	if (!strcmp(cmd, "reset-counters"))
		return cmd_reset_counters(argc, argv);

#if 0
	if (!strcmp(cmd, "assemble"))
		return cmd_assemble(argc, argv);
	if (!strcmp(cmd, "incremental"))
		return cmd_incremental(argc, argv);
	if (!strcmp(cmd, "run"))
		return cmd_run(argc, argv);
	if (!strcmp(cmd, "stop"))
		return cmd_stop(argc, argv);
#endif

	if (!strcmp(cmd, "unlock"))
		return cmd_unlock(argc, argv);
	if (!strcmp(cmd, "set-passphrase"))
		return cmd_set_passphrase(argc, argv);
	if (!strcmp(cmd, "remove-passphrase"))
		return cmd_remove_passphrase(argc, argv);

	if (!strcmp(cmd, "migrate"))
		return cmd_migrate(argc, argv);
	if (!strcmp(cmd, "migrate-superblock"))
		return cmd_migrate_superblock(argc, argv);

	if (!strcmp(cmd, "dump"))
		return cmd_dump(argc, argv);
	if (!strcmp(cmd, "list"))
		return cmd_list(argc, argv);
	if (!strcmp(cmd, "list_journal"))
		return cmd_list_journal(argc, argv);

	if (!strcmp(cmd, "setattr"))
		return cmd_setattr(argc, argv);

	if (!strcmp(cmd, "completions"))
		return cmd_completions(argc, argv);

#ifdef BCACHEFS_FUSE
	if (!strcmp(cmd, "fusemount"))
		return cmd_fusemount(argc, argv);
#endif

	if (!strcmp(cmd, "mount"))
		return cmd_mount(argc, argv);

	if (!strcmp(cmd, "--help")) {
		bcachefs_usage();
		return 0;
	}

	printf("Unknown command %s\n", cmd);
usage:
	bcachefs_usage();
	exit(EXIT_FAILURE);
}
